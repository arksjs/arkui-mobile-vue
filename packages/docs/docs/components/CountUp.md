# CountUp 数字动画

注：

- 本组件没有指定样式，默认情况下可对文字样式进行自定义。

<CodeDemo name="CountUp">

<<< @/../../demo/src/components/Show/CountUp/index.vue

</CodeDemo>

## Import

```js
import { CountUp } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type { CountUpSpeed, CountUpOnCancel, CountUpOnAnimated } from 'vfox'
```

## Props

| 属性           | 类型         | 默认值   | 必填 | 说明                                                 |
| -------------- | ------------ | -------- | ---- | ---------------------------------------------------- |
| number         | number       | 0        | 否   | 目标值，变化到该数值                                 |
| initial-number | number       | 0        | 否   | 初始值，首次动画会基于 `number` 和该值的差值来做变化 |
| speed          | CountUpSpeed | 'normal' | 否   | 可选 'normal', 'fast', 'slow'，或者固定数字的时间    |
| thousands      | boolean      | false    | 否   | 是否以千分号的形式显示，如：'1,234.56'               |
| decimal-digits | number       | 0        | 否   | 保留 `decimalDigits` 小数位数                        |

### CountUpSpeed

```ts
type CountUpSpeed = 'normal' | 'fast' | 'slow' | number
```

## Events

| 事件     | 描述                             | 回调函数参数                                  | TypeScript 函数   |
| -------- | -------------------------------- | --------------------------------------------- | ----------------- |
| animated | 动画结束后触发，主动取消也会触发 | payload: { number: number } number 为当前数值 | CountUpOnAnimated |
| cancel   | 取消成功时触发                   | payload: { number: number } number 为当前数值 | CountUpOnCancel   |

## Methods

| 方法名 | 说明             | 参数                                            |
| ------ | ---------------- | ----------------------------------------------- |
| cancel | 主动取消动画变化 | 取消成功（动画没结束前取消）会触发 cancel event |
