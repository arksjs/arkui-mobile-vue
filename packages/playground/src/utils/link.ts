export const genUnpkgLink = (
  pkg: string,
  version: string | undefined,
  path: string
) => {
  version = version ? `@${version}` : ''
  return `https://unpkg.com/${pkg}${version}${path}`
}
