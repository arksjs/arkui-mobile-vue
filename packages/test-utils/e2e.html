<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
