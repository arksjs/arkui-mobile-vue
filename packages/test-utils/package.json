{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
