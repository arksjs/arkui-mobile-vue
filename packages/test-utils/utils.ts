export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
