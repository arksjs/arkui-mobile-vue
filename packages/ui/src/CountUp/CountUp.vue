<template>
  <div class="ta-count-up">
    {{ content }}
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, watch } from 'vue'
import type { PropType } from 'vue'
import { getNumber, isNumber } from '../helpers/util'
import { thousands } from '../helpers/digital-conversion'
import type { CountUpEmits, OnCancel, Speed } from './types'
import type { PropsToEmits, VoidFnToBooleanFn } from '../helpers/types'
import { getDuration, SpeedMap } from './util'
import { useFrameTask } from '../hooks/use-frame-task'

const emitValidator: VoidFnToBooleanFn<OnCancel> = payload =>
  payload && isNumber(payload.number)

export default defineComponent({
  name: 'ta-count-up',
  props: {
    // 初始数值
    initialNumber: {
      type: [Number, String],
      default: 0
    },
    // 目标数值
    number: {
      type: [Number, String],
      default: 0
    },
    // 持续时间
    speed: {
      type: [Number, String] as PropType<Speed>,
      validator: (val: Speed) => isNumber(val) || SpeedMap.get(val) != null,
      default: 'normal'
    },
    // 是否千分位展示
    thousands: {
      type: Boolean,
      default: false
    },
    // 小数位数
    decimalDigits: {
      type: [Number, String],
      default: 0
    }
  },
  emits: {
    animated: emitValidator,
    cancel: emitValidator
  } as PropsToEmits<CountUpEmits>,
  setup(props, { emit }) {
    const content = ref('')
    let numberCache = getNumber(props.initialNumber, 0)

    const { getRunFrameTaskId, frameStart, frameStop } = useFrameTask()

    function cancel() {
      if (getRunFrameTaskId() !== null) {
        frameStop()
        emit('cancel', { number: numberCache })
      }
    }

    function update(newNumber: number) {
      cancel()

      const decimalDigits = getNumber(props.decimalDigits, 0)
      const carry = Math.pow(10, decimalDigits)
      const from = Math.round(numberCache * carry)
      const to = Math.round(newNumber * carry)

      frameStart({
        from,
        to,
        duration: getDuration(newNumber - numberCache, props.speed),
        progress: ({ current }) => {
          numberCache = parseFloat((current / carry).toFixed(decimalDigits))
          content.value = props.thousands
            ? thousands(numberCache.toFixed(decimalDigits))
            : numberCache.toFixed(decimalDigits)
        },
        complete: () => {
          emit('animated', { number: numberCache })
        }
      })
    }

    watch(
      () => props.number,
      val => {
        if (val != null && getNumber(val) !== numberCache) {
          update(getNumber(val))
        }
      },
      { immediate: true }
    )

    return {
      content,
      cancel
    }
  }
})
</script>
