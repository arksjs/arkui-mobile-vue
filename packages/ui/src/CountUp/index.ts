import { withInstall } from '../helpers/with-install'
import CountUp from './CountUp.vue'

export { CountUp }
export const install = withInstall(CountUp)
export default CountUp
