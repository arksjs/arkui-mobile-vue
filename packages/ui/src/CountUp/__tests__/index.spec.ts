import { describe, test, expect } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { CountUp } from '@/CountUp'
import { timeout } from '@arksjs/test-utils/utils'

describe('CountUp', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(CountUp, {})

    expect(wrapper.html()).toMatchSnapshot()
  })

  test.concurrent('thousands=true', async () => {
    const wrapper = shallowMount(CountUp, {
      props: {
        thousands: true,
        number: 1000,
        speed: 100
      }
    })

    await timeout(120)
    expect(wrapper.vm.content).toBe('1,000')
  })

  test.concurrent('decimalDigits=2', async () => {
    const wrapper = shallowMount(CountUp, {
      props: {
        decimalDigits: 2,
        number: 1000,
        speed: 100
      }
    })

    await timeout(120)
    expect(wrapper.vm.content).toBe('1000.00')
  })
})
