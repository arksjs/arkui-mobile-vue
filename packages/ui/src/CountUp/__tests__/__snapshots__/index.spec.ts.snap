// Vitest Snapshot v1

exports[`CountUp > snapshot 1`] = `"<div class=\\"ta-count-up\\"></div>"`;
